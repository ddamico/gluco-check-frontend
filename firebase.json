{
  "hosting": {
    "public": "build",
    "ignore": ["firebase.json", "**/.*", "**/node_modules/**"],
    "headers": [
      {
        "source": "/api/v1/validate-nightscout",
        "headers": [
          {
            "key": "Access-Control-Allow-Origin",
            "value": "*"
          },
          {
            "key": "Access-Control-Allow-Methods",
            "value": "POST,OPTIONS"
          },
          {
            "key": "Access-Control-Allow-Headers",
            "value": "Origin, Content-Type, Accept"
          }
        ]
      }
    ],
    "rewrites": [
      {
        "source": "/api/v1/validate-nightscout",
        "function": "validateNightscoutUrl"
      },
      {
        "source": "/api/v1/handle-conversation",
        "function": "conversation"
      },
      {
        "source": "**",
        "destination": "/index.html"
      }
    ]
  }
}
