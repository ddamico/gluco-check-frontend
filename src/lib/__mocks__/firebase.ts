import { BloodGlucoseUnits, DiabetesPointer } from "../enums";
import { GlucoCheckUserDocument } from "../types";

export const mockUser: Partial<firebase.User> = {
  email: "example@example.com",
  uid: "uid123",
};

export const mockUserDocument: GlucoCheckUserDocument = {
  defaultPointers: [DiabetesPointer.BloodSugar],
  glucoseUnit: BloodGlucoseUnits.mgdl,
  nightscout: {
    url: "https://example.com",
    token: "nstoken123",
  },
};
