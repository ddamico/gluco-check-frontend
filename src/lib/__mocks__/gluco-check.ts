import { BloodGlucoseUnit, DiabetesMetric } from "../enums";
import { NightscoutValidationEndpointResponse } from "../types";

export const MOCK_NSV_RESPONSE_VALID: NightscoutValidationEndpointResponse = {
  token: {
    isValid: true,
    parsed: "token123",
  },
  url: {
    isValid: true,
    parsed: "https://example.com",
    pointsToNightscout: true,
  },
  nightscout: {
    glucoseUnit: BloodGlucoseUnit.mmoll,
    minSupportedVersion: "13.0.0",
    version: "14.0.7",
  },
  discoveredMetrics: [
    DiabetesMetric.BloodSugar,
    DiabetesMetric.InsulinOnBoard,
    DiabetesMetric.CarbsOnBoard,
    DiabetesMetric.SensorAge,
    DiabetesMetric.CannulaAge,
    DiabetesMetric.PumpBattery,
  ],
};

export const MOCK_NSV_RESPONSE_NON_NS_URL = {
  url: {
    parsed: "https://example.com",
    isValid: true,
    pointsToNightscout: false,
  },
  token: {
    parsed: "",
    isValid: false,
  },
  nightscout: {
    version: "",
    glucoseUnit: "",
    minSupportedVersion: "13.0.0",
  },
  discoveredMetrics: [],
};

export const MOCK_NSV_RESPONSE_INVALID_TOKEN = {
  url: {
    parsed: "https://example.com",
    isValid: true,
    pointsToNightscout: true,
  },
  token: {
    isValid: false,
    parsed: "token1234",
  },
  nightscout: {
    minSupportedVersion: "13.0.0",
    glucoseUnit: "mmol",
    version: "14.0.7",
  },
  discoveredMetrics: ["blood sugar"],
};
