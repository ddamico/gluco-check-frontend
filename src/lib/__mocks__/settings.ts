import { BloodGlucoseUnits } from "../enums";
import { SettingsFormData } from "../types";

export const mockFormData: SettingsFormData = {
    nightscoutToken: 'I am a token',
    nightscoutUrl: 'https://example.com',
    glucoseUnit: BloodGlucoseUnits.mgdl
}
