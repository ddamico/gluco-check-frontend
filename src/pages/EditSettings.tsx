import React, { useContext } from "react";
import { useDocument } from "react-firebase-hooks/firestore";
import { firestore } from "../lib/firebase";
import { DEFAULT_USER_DOCUMENT } from "../lib/firebase-helpers";

import Container from "../components/Container";
import SettingsForm from "../components/SettingsForm";
import { SettingsFormData } from "../lib/types";
import {
  FIRESTORE_FIELD_PATH_GLUCOSE_UNITS,
  FIRESTORE_FIELD_PATH_NIGHTSCOUT_TOKEN,
  FIRESTORE_FIELD_PATH_NIGHTSCOUT_URL,
} from "../lib/constants";
import { userSettingsFormDataToUserSettingsDocument } from "../lib/transform";
import { FirebaseUserDocumentContext } from "../App";
import { useTranslation } from "react-i18next";

export const returnHandleSettingsSave = (userDocumentPath: string) => {
  return async (data: SettingsFormData) => {
    const formDataAsDocument = userSettingsFormDataToUserSettingsDocument(data);
    try {
      await firestore.doc(userDocumentPath).set({
        ...DEFAULT_USER_DOCUMENT,
        ...formDataAsDocument,
      });
    } catch (e) {
      throw new Error(`error saving document ${e.message}`);
    }
  };
};

// @TODO: add success and error states to this, and set them in the handler?
export default function EditSettings() {
  const userDocumentPath = useContext(FirebaseUserDocumentContext);
  const userDocumentReference = firestore.doc(userDocumentPath);

  const [document, loading, error] = useDocument(userDocumentReference);

  const { t } = useTranslation();

  const nightscoutUrl =
    document?.get(FIRESTORE_FIELD_PATH_NIGHTSCOUT_URL) ?? "";
  const nightscoutToken =
    document?.get(FIRESTORE_FIELD_PATH_NIGHTSCOUT_TOKEN) ?? "";
  const glucoseUnit = document?.get(FIRESTORE_FIELD_PATH_GLUCOSE_UNITS) ?? "";

  return (
    <Container>
      <h2>{t("settings.title")}</h2>
      {loading && <>Loading stuff!</>}
      {error && <>Something has gone horribly wrong: {error.message}</>}
      {document && (
        <SettingsForm
          nightscoutUrl={nightscoutUrl}
          nightscoutToken={nightscoutToken}
          glucoseUnit={glucoseUnit}
          onSubmit={returnHandleSettingsSave(userDocumentPath)}
        />
      )}
    </Container>
  );
}
