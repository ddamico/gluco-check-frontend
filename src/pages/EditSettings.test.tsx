import React from "react";
import { cleanup, render } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";
import * as firestoreHooks from "react-firebase-hooks/firestore";
import EditSettings, { returnHandleSettingsSave } from "./EditSettings";
import { mockFirebase } from "firestore-jest-mock";
import { mockUser, mockUserDocument } from "../lib/__mocks__/firebase";

expect.extend(toHaveNoViolations);

afterEach(() => {
  cleanup();
  jest.clearAllMocks();
});

beforeEach(() => {
  mockFirebase({
    database: {
      users: {
        "alf@melmac.com": mockUserDocument,
      },
    },
  });
});

describe("EditSettings", () => {
  it("Renders the component when loading, with no axe violations", async () => {
    // @ts-ignore
    firestoreHooks.useDocument = jest
      .fn()
      .mockReturnValue([undefined, true, undefined]);

    // @ts-ignore
    const { container } = render(
      <EditSettings userDocumentPath="users/alf@melmac.com" />
    );
    expect(container.firstChild).toMatchSnapshot();

    expect(await axe(container)).toHaveNoViolations();
  });

  it("Renders the component when in error state, with no axe violations", async () => {});

  it("Renders the component with data, with no axe violations", async () => {
    const mockDocument = {
      get: jest.fn().mockReturnValue("mock data"),
    };
  });
});

describe("returnHandleSettingsSave", () => {
  it("returns a submit handler", () => {
    // @ts-ignore
    const handler = returnHandleSettingsSave(mockUser);
    expect(handler).toBeInstanceOf(Function);
  });
});
