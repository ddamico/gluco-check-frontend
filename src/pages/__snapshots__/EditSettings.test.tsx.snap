// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSettings Renders the component when loading, with no axe violations 1`] = `
<div
  class="MuiContainer-root makeStyles-container-1 MuiContainer-maxWidthLg"
>
  <h2
    class="MuiTypography-root MuiTypography-h6"
  >
    settings.title
  </h2>
  <p>
    settings.content
  </p>
  <form
    class="makeStyles-form-3"
    data-testid="settings-form"
  >
    <fieldset
      class="MuiFormControl-root"
      data-testid="settings-form-fieldset-metrics"
    >
      <legend
        class="MuiFormLabel-root"
      >
        settings.form.labels.defaultMetrics
      </legend>
      <div
        class="MuiFormGroup-root makeStyles-checkboxArray-4 MuiFormGroup-row"
      >
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                class="PrivateSwitchBase-input-8"
                data-indeterminate="false"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            everything
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                class="PrivateSwitchBase-input-8"
                data-indeterminate="false"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            blood sugar
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                class="PrivateSwitchBase-input-8"
                data-indeterminate="false"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            insulin on board
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                class="PrivateSwitchBase-input-8"
                data-indeterminate="false"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            carbs on board
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                class="PrivateSwitchBase-input-8"
                data-indeterminate="false"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            sensor age
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                class="PrivateSwitchBase-input-8"
                data-indeterminate="false"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            cannula age
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-5 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                class="PrivateSwitchBase-input-8"
                data-indeterminate="false"
                type="checkbox"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            pump battery
          </span>
        </label>
      </div>
      <p
        class="MuiFormHelperText-root"
      >
        settings.form.helperText.defaultMetrics
      </p>
    </fieldset>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink="true"
        for="settings-form-field-url"
        id="settings-form-field-url-label"
      >
        settings.form.labels.nightscoutUrl
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-describedby="settings-form-field-url-helper-text"
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input"
          data-testid="settings-form-field-url"
          id="settings-form-field-url"
          name="nightscoutUrl"
          type="text"
          value="a value"
        />
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-filled"
        id="settings-form-field-url-helper-text"
      >
        settings.form.helperText.nightscoutUrl
      </p>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink="true"
        for="settings-form-field-token"
        id="settings-form-field-token-label"
      >
        settings.form.labels.nightscoutToken
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
      >
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionStart"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
            />
          </svg>
        </div>
        <input
          aria-describedby="settings-form-field-token-helper-text"
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
          data-testid="settings-form-field-token"
          id="settings-form-field-token"
          name="nightscoutToken"
          type="text"
          value="a value"
        />
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-filled"
        id="settings-form-field-token-helper-text"
      >
        <button
          class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiLink-button MuiTypography-colorPrimary"
          type="button"
        >
          settings.form.helperText.nightscoutToken
        </button>
      </p>
    </div>
    <div
      class="MuiFormControl-root MaterialSelect MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink="true"
      >
        settings.form.labels.glucoseUnits
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="settings-form-field-bg"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="settings-form-field-bg"
          role="button"
          tabindex="0"
        >
          <span>
            ​
          </span>
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          data-testid="settings-form-field-bg"
          name="glucoseUnit"
          tabindex="-1"
          value="a value"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthLg"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        data-testid="settings-form-submit"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          settings.form.submitButton
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`EditSettings Renders the component when loading, with no axe violations 2`] = `
<div
  class="MuiContainer-root makeStyles-container-9 MuiContainer-maxWidthLg"
>
  <h2
    class="MuiTypography-root MuiTypography-h6"
  >
    settings.title
  </h2>
  <p>
    settings.content
  </p>
  status.general.loading
</div>
`;

exports[`EditSettings Renders the component with error, with no axe violations 1`] = `
<div
  class="MuiContainer-root makeStyles-container-10 MuiContainer-maxWidthLg"
>
  <h2
    class="MuiTypography-root MuiTypography-h6"
  >
    settings.title
  </h2>
  <p>
    settings.content
  </p>
  status.general.error
  : 
  you blew it
</div>
`;
